### PR TITLE
fix attention

### PR DESCRIPTION
**Pull request by alexholdenmiller**
#596

### CODE REVIEW

**Comment by alexholdenmiller** [comment]
added post-attention as well, which appears to do better more frequently. (instead of modifying the embeddings of the inputs to the rnn, it modifies the outputs of the rnn before they are fed to the output layer, as described in )

with a few runs each (so there's variability in sampling...), babi task 19, with validation every 30 secs and an impatience of 10 (so not very patient).

no attention: 9% valid, 8% test || 17% valid, 14% test || 9% valid, 8% test

general pre-attention: 12% valid, 10% test || 13% valid, 12% test || 10% valid, 9% test 
general post-attention: 77% valid, 74% test || 80% valid, 81% test || 11% valid, 9% test || 7% valid, 8% test

dot pre-attention: 25% valid, 26% test || 12% valid, 11% test || 12% valid, 9% test
dot post-attention: 52% valid, 53% test || 12% valid, 10% test || 11% valid, 11% test || 10% valid, 11% test || 12% valid, 11% test || 41% valid, 41% test

concat pre-attention: 10% valid, 8% test || 12% valid, 10% test || 9% valid, 10% test || 9% valid, 8% test
concat post-attention: 14% valid, 13% test || 39% valid, 38% test || 12% valid, 9% test

in hindsight, there's more variability here than I expected, but post-attention appears to be the most reliable for getting above random chance. running on personachat now.

File: parlai/agents/seq2seq/modules.py [hunk 1]
```diff
@@ -434,15 +442,23 @@ def forward(self, input):
 
 class AttentionLayer(nn.Module):
     def __init__(self, attn_type, hidden_size, emb_size, bidirectional=False,
-                 attn_length=-1):
+                 attn_length=-1, attn_time='pre'):
         super().__init__()
         self.attention = attn_type
 
         if self.attention != 'none':
             hsz = hidden_size
```
Comment (emilydinan) [inline parlai/agents/seq2seq/modules.py]: why do we rename this variable here?

Comment (alexholdenmiller) [reply parlai/agents/seq2seq/modules.py]: it made the lines below a lot shorter :)

**Comment by alexholdenmiller** [comment]
quick note: after ~15 min, dot post-attention is producing pretty decent looking personachat sentences already, with no attetion, general post, and concat post all looking pretty poor.

**Comment by alexholdenmiller** [comment]
general post looks better on the 4th validation, and oops meant 30 min not 15

**Pull request merged**